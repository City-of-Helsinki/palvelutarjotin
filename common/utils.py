from datetime import datetime

import graphene
from django.conf import settings
from django.core.exceptions import PermissionDenied
from django.db import transaction
from django.utils import timezone
from graphene import Node
from graphql_relay import from_global_id

from palvelutarjotin import __version__
from palvelutarjotin.exceptions import (
    DataValidationError,
    IncorrectGlobalIdError,
    ObjectDoesNotExistError,
)
from palvelutarjotin.settings import REVISION

LanguageEnum = graphene.Enum(
    "Language", [(l[0].upper(), l[0]) for l in settings.LANGUAGES]
)

LINKED_EVENT_DATE_FORMAT = "%Y-%m-%d"


def format_linked_event_date(datetime_obj):
    return datetime_obj.strftime(LINKED_EVENT_DATE_FORMAT)


def update_object(obj, data):
    if not data:
        return
    for k, v in data.items():
        if v is None and not obj.__class__._meta.get_field(k).null:
            raise DataValidationError(f"{k} cannot be null.")
        setattr(obj, k, v)
    obj.save()


@transaction.atomic
def update_object_with_translations(model, model_data):
    translations_input = model_data.pop("translations", None)
    if translations_input:
        model.create_or_update_translations(translations_input)
    update_object(model, model_data)


def get_api_version():
    return " | ".join((__version__, REVISION.decode("utf-8")))


def get_node_id_from_global_id(global_id, node_name):
    name, id = from_global_id(global_id)
    if name != node_name:
        raise IncorrectGlobalIdError("Node type does not match")
    return id


def convert_to_localtime_tz(value):
    dt = datetime.combine(datetime.now().date(), value)
    if timezone.is_naive(value):
        # Auto add local timezone to naive time
        return timezone.make_aware(dt).timetz()
    else:
        return timezone.localtime(dt).timetz()


def get_obj_from_global_id(info, global_id, expected_obj_type):
    obj = Node.get_node_from_global_id(info, global_id)
    if not obj or type(obj) != expected_obj_type:
        raise ObjectDoesNotExistError(
            f"{expected_obj_type.__name__} matching query does not exist."
        )
    return obj


def get_editable_obj_from_global_id(info, global_id, expected_obj_type):
    obj = get_obj_from_global_id(info, global_id, expected_obj_type)
    try:
        if obj.is_editable_by_user(info.context.user):
            return obj
    except AttributeError:
        raise TypeError(
            f"{obj.__class__.__name__} model does not implement is_editable_by_user()."
        )
    raise PermissionDenied(
        f"User does not have permission to edit this {expected_obj_type.__name__}"
    )
