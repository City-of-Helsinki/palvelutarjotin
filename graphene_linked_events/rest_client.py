import requests


class LinkedEventsApiClient(object):

    CONNECTION_TIMEOUT = 5

    def __init__(self, config) -> None:
        self.root = config["ROOT"]
        self.api_key = config["API_KEY"]
        super().__init__()

    def get_actions(self, resource=None):
        url = self.root + resource
        return {
            "list": {"method": "GET", "url": url + "/"},
            "create": {"method": "POST", "url": url + "/"},
            "retrieve": {"method": "GET", "url": url + "/{}/"},
            "update": {"method": "PUT", "url": url + "/{}/"},
            "delete": {"method": "DELETE", "url": url + "/{}/"},
            "search": {"method": "GET", "url": self.root + "search/"},
            "upload": {"method": "POST", "url": url + "/"},
        }

    def retrieve(self, resource, id, params=None, is_staff=False):
        actions = self.get_actions(resource)
        formatted_params = self.convert_to_string_param(params)
        if is_staff:
            headers = {"apikey": self.api_key, "Cache-Control": "no-cache"}
            cookies = dict(nocache="meow")
            return requests.request(
                actions["retrieve"]["method"],
                actions["retrieve"]["url"].format(id),
                params=formatted_params,
                headers=headers,
                cookies=cookies,
                timeout=self.CONNECTION_TIMEOUT,
            )
        return requests.request(
            actions["retrieve"]["method"],
            actions["retrieve"]["url"].format(id),
            params=formatted_params,
            timeout=self.CONNECTION_TIMEOUT,
        )

    def list(self, resource, filter_list=None, is_staff=False):
        actions = self.get_actions(resource)
        filter_params = self.convert_to_string_param(filter_list)
        if is_staff:
            headers = {"apikey": self.api_key, "Cache-Control": "no-cache"}
            cookies = dict(nocache="meow")
            return requests.request(
                actions["list"]["method"],
                actions["list"]["url"],
                params=filter_params,
                headers=headers,
                cookies=cookies,
                timeout=self.CONNECTION_TIMEOUT,
            )
        return requests.request(
            actions["list"]["method"],
            actions["list"]["url"],
            params=filter_params,
            timeout=self.CONNECTION_TIMEOUT,
        )

    def create(self, resource, body):
        actions = self.get_actions(resource)
        headers = {"apikey": self.api_key, "Content-Type": "application/json"}
        return requests.request(
            actions["create"]["method"],
            actions["create"]["url"],
            data=body,
            headers=headers,
            timeout=self.CONNECTION_TIMEOUT,
        )

    def update(self, resource, id, body):
        actions = self.get_actions(resource)
        headers = {"apikey": self.api_key, "Content-Type": "application/json"}
        return requests.request(
            actions["update"]["method"],
            actions["update"]["url"].format(id),
            data=body,
            headers=headers,
            timeout=self.CONNECTION_TIMEOUT,
        )

    def delete(self, resource, id):
        actions = self.get_actions(resource)
        headers = {"apikey": self.api_key, "Content-Type": "application/json"}
        return requests.request(
            actions["delete"]["method"],
            actions["delete"]["url"].format(id),
            headers=headers,
            timeout=self.CONNECTION_TIMEOUT,
        )

    # Special action to full-text search generic resources
    def search(self, params):
        search_params = self.convert_to_string_param(params)
        action = self.get_actions()["search"]
        response = requests.request(
            action["method"],
            action["url"],
            params=search_params,
            timeout=self.CONNECTION_TIMEOUT,
        )
        return response

    def upload(self, resource, body, files):
        action = self.get_actions(resource)["upload"]
        headers = {"apikey": self.api_key}
        return requests.request(
            action["method"],
            action["url"],
            data=body,
            files=files,
            headers=headers,
            timeout=self.CONNECTION_TIMEOUT,
        )

    @staticmethod
    def convert_to_string_param(params):
        if not params:
            return None
        for k, v in params.items():
            if type(v) == list:
                list_to_string = ",".join(v)
                params[k] = list_to_string
        return params
