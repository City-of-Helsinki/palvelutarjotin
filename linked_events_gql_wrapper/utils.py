# https://stackoverflow.com/questions/6578986/how-to-convert-json-data
# -into-a-python-object/15882054#15882054
import json
from collections import namedtuple


def format_response(response):
    # Some fields from api have @prefix that need to be converted
    return response.content.decode("utf-8").replace("@", "internal_")


def _json_object_hook(d):
    return namedtuple("X", d.keys())(*d.values())


def json2obj(data):
    return json.loads(data, object_hook=_json_object_hook)
