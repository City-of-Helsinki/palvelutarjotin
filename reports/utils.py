from math import asin, cos, radians, sin, sqrt
from typing import List, Optional

LINKED_EVENTS_GEOS_PROJECTION_SRID = 3067

"""
FIXME: get_distance would be better alternative to haversine(),
since it calculates the 3d distance.
Can the 3d distance really be twice the 2d distance or why is the number so much bigger?
"""
# from django.contrib.gis.measure import Distance
# from django.contrib.gis.geos import Point
# def get_distance(lon1: float, lat1: float, lon2: float, lat2: float) -> float:
#     """Get a 3D distance in meters between the 2 coordinates."""
#     start = Point(lon1, lat1, srid=LINKED_EVENTS_GEOS_PROJECTION_SRID)
#     end = Point(lon2, lat2, srid=LINKED_EVENTS_GEOS_PROJECTION_SRID)
#     return Distance(km=start.distance(end) * 100).km


def haversine(lon1: float, lat1: float, lon2: float, lat2: float) -> float:
    """
    Calculate the great circle distance in kilometers between two points
    on the earth (specified in decimal degrees)
    HAversine formula: https://en.wikipedia.org/wiki/Haversine_formula.
    """
    # convert decimal degrees to radians
    lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])

    # haversine formula
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = sin(dlat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(dlon / 2) ** 2
    c = 2 * asin(sqrt(a))
    r = 6371  # Radius of earth in kilometers. Determines return value units.
    return c * r


def get_event_provider(event_json: dict, language: str = "fi") -> Optional[str]:
    """Return a provider name in preferred language from given event json.

    Args:
        event_json (dict): [description]
        language (str, optional): [description]. Defaults to "fi".

    Returns:
        Optional[str]: [description]
    """
    return event_json["provider"]["fi"] if event_json["provider"] else None


def get_event_keywords(event_json: dict, language: str = "fi") -> List[str]:
    """Return event keywords in a format that is supported by EnrolmentReport model.

    Args:
        event_json (dict): event json fetched from LinkedEvents API
        language (str, optional): [description]. Defaults to "fi".

    Returns:
        List[str]: First argument is an index, second is a translation or an empty list
    """
    return (
        [
            (kw["id"], kw.get("name", {language: ""})[language]) if kw else ""
            for kw in event_json["keywords"]
        ]
        if event_json["keywords"]
        else []
    )
