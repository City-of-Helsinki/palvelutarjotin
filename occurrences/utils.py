from django_ilmoitin.utils import send_notification


def send_event_notifications_to_contact_person(
    occurrence, study_group, notification_type, **kwargs
):
    context = {
        "occurrence": occurrence,
        "study_group": study_group,
        **kwargs,
    }
    # TODO: Send notification based on user language
    send_notification(
        study_group.person.email_address, notification_type, context=context,
    )
