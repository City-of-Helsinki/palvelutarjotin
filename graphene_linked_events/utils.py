# https://stackoverflow.com/questions/6578986/how-to-convert-json-data
# -into-a-python-object/15882054#15882054
import json
from collections import namedtuple
from django.conf import settings
from geopy import distance as geopy_distance
from geopy import Point
from typing import List

from graphene_linked_events.rest_client import LinkedEventsApiClient
from palvelutarjotin.exceptions import ApiBadRequestError, ObjectDoesNotExistError

api_client = LinkedEventsApiClient(config=settings.LINKED_EVENTS_API_CONFIG)


def format_response(response):
    """Remove @ sign from API response JSON keys."""

    def format_values(value):
        if isinstance(value, list):
            return [format_values(v) for v in value]
        elif isinstance(value, dict):
            return {
                key.replace("@", "internal_"): format_values(value)
                for key, value in value.items()
            }
        return value

    formatted = format_values(json.loads(response.text))
    return json.dumps(formatted)


def json_object_hook(d):
    return namedtuple("X", d.keys())(*d.values())


def json2obj(data):
    return json.loads(data, object_hook=json_object_hook)


def format_request(request):
    # TODO: Find better way to replace internal_id key
    return json.dumps(request).replace("internal_", "@")


def retrieve_linked_events_data(resource, resource_id, params=None, is_staff=False):
    response = api_client.retrieve(
        resource, resource_id, params=params, is_staff=is_staff
    )

    if response.status_code == 400:
        raise ApiBadRequestError("Could not establish a connection to the API.")

    if response.status_code == 404:
        raise ObjectDoesNotExistError("Could not find the event from the API.")

    # Raise any other errors
    response.raise_for_status()

    return json2obj(format_response(response))


def get_keyword_set_by_id(keyword_set_id):
    params = {"include": "keywords"}
    return retrieve_linked_events_data(
        "keyword_set",
        keyword_set_id,
        params=params,
    )


def bbox_for_coordinates(
    longitude: float, latitude: float, distance: float
) -> List[float]:
    """
    Calculate bounding box for the given coordinates so that the
    bounding box corner is at the given distance (in km).
    """
    sw_bearing = 225
    ne_bearing = 45
    d = geopy_distance.distance(kilometers=distance)

    starting_point = Point(latitude=latitude, longitude=longitude)
    sw_point = d.destination(starting_point, bearing=sw_bearing)
    ne_point = d.destination(starting_point, bearing=ne_bearing)

    return [
        sw_point.longitude,
        sw_point.latitude,
        ne_point.longitude,
        ne_point.latitude,
    ]
