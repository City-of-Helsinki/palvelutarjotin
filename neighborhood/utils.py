import json
from collections import namedtuple

from django.conf import settings
from neighborhood.rest_client import NeighborhoodApiClient

api_client = NeighborhoodApiClient(config=settings.NEIGHBORHOOD_API_CONFIG)


def format_response(response):
    return response.text


def json_object_hook(d):
    return namedtuple("X", d.keys())(*d.values())


def json2obj(data):
    return json.loads(data, object_hook=json_object_hook)


def format_request(request):
    return json.dumps(request)
