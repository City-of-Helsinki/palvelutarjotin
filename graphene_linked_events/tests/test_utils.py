import math

from graphene_linked_events.utils import bbox_for_coordinates


def test_bbox_for_coordinates():
    """
    Bounding box for Katri Valan puisto with a distance of 3 km from center to corner.
    """
    bbox = bbox_for_coordinates(24.963692, 60.186687, 3)

    assert math.isclose(bbox[0], 24.925482025464998)
    assert math.isclose(bbox[1], 60.16764173276239)
    assert math.isclose(bbox[2], 25.00194624440269)
    assert math.isclose(bbox[3], 60.20572118886215)
