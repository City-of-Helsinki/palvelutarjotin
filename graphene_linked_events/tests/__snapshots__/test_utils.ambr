# serializer version: 1
# name: test_format_response
  dict({
    'data': list([
      dict({
        'email': 'email1@example.com',
        'emails': list([
          'email2@example.com',
          'email3@example.com',
        ]),
        'internal_id': 123,
      }),
      dict({
        'internal_key': '@@@',
        'internal_keys': list([
          dict({
            'internal_key': 'value',
          }),
          dict({
            'key': '@value',
          }),
        ]),
      }),
    ]),
  })
# ---
